code/iaas/logic/src/main/java/io/cattle/platform/process/containerevent/ContainerEventCreate.java

```diff
@@ -61,76 +68,109 @@
     AccountDao accountDao;
 
     @Inject
+    InstanceDao instanceDao;
+
+    @Inject
     LockManager lockManager;
 
     @SuppressWarnings("unchecked")
     @Override
     public HandlerResult handle(ProcessState state, ProcessInstance process) {
-        if ( !MANAGE_NONRANCHER_CONTAINERS.get() ) {
+        if (!MANAGE_NONRANCHER_CONTAINERS.get()) {
             return null;
         }
 
-        ContainerEvent event = (ContainerEvent)state.getResource();
-        Map<String, Object> inspect = (Map<String, Object>)DataUtils.getFields(event).get(FIELD_DOCKER_INSPECT);
+        final ContainerEvent event = (ContainerEvent)state.getResource();
 
-        String rancherUuid = getRancherUuidLabel(inspect);
-        if ( rancherUuid != null ) {
-            return null;
-        }
+        HandlerResult result = lockManager.lock(new ContainerEventInstanceLock(event.getAccountId(), event.getExternalId()), new LockCallback<HandlerResult>() {
+            @Override
+            public HandlerResult doWithLock() {
+                Map<String, Object> inspect = (Map<String, Object>)DataUtils.getFields(event).get(FIELD_DOCKER_INSPECT);
+                String rancherUuid = getRancherUuidLabel(inspect);
+                Instance instance = instanceDao.getInstanceByExternalIdOrUuid(event.getAccountId(), event.getExternalId(), rancherUuid);
+                if (instance != null && StringUtils.isNotEmpty(instance.getSystemContainer())) {
+                    // System containers are not managed by contaienr events
+                    return null;
+                }
 
-        scheduleInstance(event, inspect);
+                try {
+                    String status = event.getExternalStatus();
+                    if (status.equals(EVENT_CREATE)) {
+                        scheduleInstance(event, instance, inspect);
+                        return null;
+                    }
 
-        return null;
+                    if (instance == null) {
+                        return null;
+                    }
+
+                    String state = ObjectUtils.getState(instance);
+                    if (EVENT_START.equals(status)) {
+                        if (STATE_RUNNING.equals(state) || STATE_STARTING.equals(state))
+                            return null;
+
+                        objectProcessManager.scheduleProcessInstance(PROCESS_START, instance, makeData());
+                    } else if (EVENT_STOP.equals(status) || EVENT_DIE.equals(status)) {
+                        if (STATE_STOPPED.equals(state) || STATE_STOPPING.equals(state))
+                            return null;
+
+                        objectProcessManager.scheduleProcessInstance(PROCESS_STOP, instance, makeData());
+                    } else if (EVENT_DESTROY.equals(status)) {
+                        if (REMOVED.equals(state) || REMOVING.equals(state) || PURGED.equals(state) || PURGING.equals(state))
```

Comment:
Instead of checking for state being Purged or purging, we can also check for removed flag not being null. But its up to you, minor thing